tests/functional/retry/test_retry.py

```diff
@@ -0,0 +1,171 @@
+        }
+
+        assert {n.node.name: n.status for n in results.results} == expected_statuses
+
+        # No failures in previous run, nothing to retry
+        results = run_dbt(["retry"])
+        expected_statuses = {}
+        assert {n.node.name: n.status for n in results.results} == expected_statuses
+
+        write_file(models__sample_model, "models", "sample_model.sql")
+
+    def test_warn_error(self, project):
+        # Regular build
+        results = run_dbt(["--warn-error", "build", "--select", "second_model"], expect_pass=False)
+
+        expected_statuses = {
+            "second_model": RunStatus.Success,
+            "accepted_values_second_model_bar__False__3": TestStatus.Fail,
+        }
+
+        assert {n.node.name: n.status for n in results.results} == expected_statuses
+
+        # Retry regular, should pass
+        run_dbt(["retry"])
+
+        # Retry with --warn-error, should fail
+        run_dbt(["--warn-error", "retry"], expect_pass=False)
+
+    def test_custom_target(self, project):
+        run_dbt(["build", "--select", "second_model"])
+        run_dbt(
+            ["build", "--select", "sample_model", "--target-path", "target2"], expect_pass=False
+        )
+
+        # Regular retry
+        results = run_dbt(["retry"])
+        expected_statuses = {"accepted_values_second_model_bar__False__3": TestStatus.Warn}
+        assert {n.node.name: n.status for n in results.results} == expected_statuses
+
+        # Retry with custom target
+        fixed_sql = "select 1 as id, 1 as foo"
+        write_file(fixed_sql, "models", "sample_model.sql")
+
+        results = run_dbt(["retry", "--state", "target2"])
+        expected_statuses = {
+            "sample_model": RunStatus.Success,
+            "accepted_values_sample_model_foo__False__3": TestStatus.Warn,
+        }
+
+        assert {n.node.name: n.status for n in results.results} == expected_statuses
+
+        write_file(models__sample_model, "models", "sample_model.sql")
+
+    def test_run_operation(self, project):
+        results = run_dbt(
+            ["run-operation", "alter_timezone", "--args", "{timezone: abc}"], expect_pass=False
+        )
+
+        expected_statuses = {
+            "operation.test.alter_timezone": RunStatus.Error,
+        }
+
+        assert {n.unique_id: n.status for n in results.results} == expected_statuses
+
+        results = run_dbt(["retry"], expect_pass=False)
+        assert {n.unique_id: n.status for n in results.results} == expected_statuses
+
+    def test_fail_fast(self, project):
+        result = run_dbt(["--warn-error", "build", "--fail-fast"], expect_pass=False)
+
+        assert result.status == RunStatus.Error
+        assert result.node.name == "sample_model"
+
+        results = run_dbt(["retry"], expect_pass=False)
+
+        assert len(results.results) == 1
+        assert results.results[0].status == RunStatus.Error
+        assert results.results[0].node.name == "sample_model"
```

Comment:
From discussing https://github.com/dbt-labs/dbt-core/pull/7763/files#r1218248760 live: @aranke added a test case for this. Current behavior of `--fail-fast` is to only return **exactly one** result, and only for the node that encountered the error (which triggered the fast failure). That's true no matter how many nodes are being run, and even if previous nodes had already succeeded.

Let's consider this out of scope for this PR (adding the `retry` command), but open a new issue that we may want to tackle (if we have time) in this epic